File: app/src/main/kotlin/com/goodgym/core/ExerciseCounter.kt
package com.goodgym.core

import kotlin.math.atan2
import kotlin.math.sqrt

/**
 * 运动计数器 - 翻译自桌面版 [exercise_counters.py]
 *
 * 核心算法:
 *   1. 三点夹角 = atan2(cross, dot) 计算关节角度
 *   2. 滑动窗口平滑: 中位数 + 2σ 离群点过滤
 *   3. up/down 阶段切换 + 0.5s 最小 rep 间隔
 *   4. 腿部动作独立计数: 左右腿各自维护 stage 状态
 *
 * 状态机:
 *   down -> up: 关节伸过 up_angle 阈值
 *   up -> down: 关节弯过 down_angle 阈值
 *   一次完整 down -> up -> down 算 1 rep (或方向相反)
 */
class ExerciseCounter(
    private val smoothingWindow: Int = 5
) {
    enum class Stage { NONE, UP, DOWN }

    var count: Int = 0
        private set

    private var stage: Stage = Stage.NONE
    private var leftLegStage: Stage = Stage.NONE
    private var rightLegStage: Stage = Stage.NONE

    private val angleHistory = ArrayDeque<Float>()
    private var lastCountTimeMs: Long = 0

    companion object {
        private const val MIN_REP_INTERVAL_MS = 500L
    }

    /**
     * 处理一帧关键点, 返回当前平滑角度 (度数, 可能为 null)
     *
     * [keypoints] 长度 34 的 FloatArray (17 个点 × 2 坐标)
     */
    fun update(keypoints: FloatArray, config: ExerciseConfig): Float? {
        val kp = config.keypoints

        // 计算左右两侧角度
        val leftAngle = calculateAngle(
            keypoints, kp.left[0], kp.left[1], kp.left[2]
        ) ?: return null
        val rightAngle = calculateAngle(
            keypoints, kp.right[0], kp.right[1], kp.right[2]
        ) ?: return null

        // 腿部动作: 双腿独立计数
        if (config.is_leg_exercise) {
            return countLegExercise(leftAngle, rightAngle, config)
        }

        // 其他动作: 左右平均
        val avg = (leftAngle + rightAngle) / 2
        val smoothed = smoothAngle(avg) ?: return null

        val upThr = config.up_angle
        val downThr = config.down_angle
        val direction = getAngleDirection(config)

        if (direction == "decreasing") {
            when {
                smoothed > downThr -> stage = Stage.DOWN
                smoothed < upThr && stage == Stage.DOWN && checkRepTiming() -> {
                    stage = Stage.UP
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                }
            }
        } else {
            when {
                smoothed > upThr -> stage = Stage.UP
                smoothed < downThr && stage == Stage.UP && checkRepTiming() -> {
                    stage = Stage.DOWN
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                }
            }
        }

        return smoothed
    }

    private fun countLegExercise(
        leftAngle: Float, rightAngle: Float, config: ExerciseConfig
    ): Float {
        val upThr = config.up_angle
        val downThr = config.down_angle
        val direction = getAngleDirection(config)

        if (checkRepTiming()) {
            // 左腿
            if (direction == "decreasing") {
                val leftReady = leftAngle > downThr
                val leftCounted = leftAngle < upThr && leftLegStage == Stage.DOWN
                if (leftReady) {
                    leftLegStage = Stage.DOWN
                } else if (leftCounted) {
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                    leftLegStage = Stage.UP
                }
            } else {
                val leftReady = leftAngle > upThr
                val leftCounted = leftAngle < downThr && leftLegStage == Stage.UP
                if (leftReady) {
                    leftLegStage = Stage.UP
                } else if (leftCounted) {
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                    leftLegStage = Stage.DOWN
                }
            }

            // 右腿
            if (direction == "decreasing") {
                val rightReady = rightAngle > downThr
                val rightCounted = rightAngle < upThr && rightLegStage == Stage.DOWN
                if (rightReady) {
                    rightLegStage = Stage.DOWN
                } else if (rightCounted) {
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                    rightLegStage = Stage.UP
                }
            } else {
                val rightReady = rightAngle > upThr
                val rightCounted = rightAngle < downThr && rightLegStage == Stage.UP
                if (rightReady) {
                    rightLegStage = Stage.UP
                } else if (rightCounted) {
                    count++
                    lastCountTimeMs = System.currentTimeMillis()
                    rightLegStage = Stage.DOWN
                }
            }
        }

        return (leftAngle + rightAngle) / 2
    }

    private fun getAngleDirection(config: ExerciseConfig): String {
        config.angle_direction?.let { if (it == "increasing" || it == "decreasing") return it }
        return if (config.up_angle < config.down_angle) "decreasing" else "increasing"
    }

    private fun checkRepTiming(): Boolean {
        val now = System.currentTimeMillis()
        return now - lastCountTimeMs >= MIN_REP_INTERVAL_MS
    }

    private fun smoothAngle(angle: Float): Float? {
        angleHistory.addLast(angle)
        if (angleHistory.size > smoothingWindow) angleHistory.removeFirst()
        if (angleHistory.size < 3) return angle

        // 中位数
        val sorted = angleHistory.sorted()
        val median = sorted[sorted.size / 2]
        // 标准差
        val mean = angleHistory.toFloatArray().sum() / angleHistory.size
        val variance = angleHistory.toFloatArray().map { (it - mean) * (it - mean) }.sum() / angleHistory.size
        val std = sqrt(variance)
        // 过滤 2σ 离群点
        val filtered = angleHistory.filter { kotlin.math.abs(it - median) <= 2 * std }
        return if (filtered.isNotEmpty()) filtered.toFloatArray().sum() / filtered.size else angle
    }

    /**
     * 计算三点夹角 - 端点 a, 关节 b, 端点 c
     *
     * 返回 (0, 180] 度数, 任意点为 (0, 0) 视为无效返回 null
     */
    private fun calculateAngle(
        kps: FloatArray, aIdx: Int, bIdx: Int, cIdx: Int
    ): Float? {
        val ax = kps[aIdx * 2]; val ay = kps[aIdx * 2 + 1]
        val bx = kps[bIdx * 2]; val by = kps[bIdx * 2 + 1]
        val cx = kps[cIdx * 2]; val cy = kps[cIdx * 2 + 1]
        // 任意一点为 (0, 0) 视为无效 (低置信度被过滤)
        if (ax == 0f && ay == 0f) return null
        if (bx == 0f && by == 0f) return null
        if (cx == 0f && cy == 0f) return null

        val bax = ax - bx; val bay = ay - by
        val bcx = cx - bx; val bcy = cy - by
        val dot = bax * bcx + bay * bcy
        val cross = bax * bcy - bay * bcx
        var ang = Math.toDegrees(atan2(cross.toDouble(), dot.toDouble())).toFloat()
        if (ang < 0) ang += 180f
        return ang
    }

    fun reset() {
        count = 0
        stage = Stage.NONE
        leftLegStage = Stage.NONE
        rightLegStage = Stage.NONE
        angleHistory.clear()
        lastCountTimeMs = 0
    }
}

File: lib/core/exercise_counter.dart
/// 运动计数核心 - 翻译自桌面版 [exercise_counters.py](file:///f:/git/Good-GYM/exercise_counters.py)
///
/// 逻辑保持与桌面版一致:
/// - 三点夹角 (COCO 17 关键点)
/// - 滑动窗口平滑 (中位数 + 2σ 离群点过滤)
/// - 状态机计数 (up/down 阶段切换)
/// - 腿部动作左右腿独立计数
/// - 最小 rep 间隔 (0.5s) 防误触
library;

import 'dart:math' as math;
import 'dart:typed_data';

import 'exercise_config.dart';

/// 关键点坐标, 形如 [[x0, y0], [x1, y1], ...] 共 17 个
typedef Keypoints = List<Float64List>;

/// 计数阶段
enum Stage { up, down, none }

class ExerciseCounter {
  /// 当前累计次数
  int count = 0;

  /// 当前阶段 (用于状态机判断)
  Stage _stage = Stage.none;

  /// 腿部动作的左右腿独立阶段
  Stage _leftLegStage = Stage.none;
  Stage _rightLegStage = Stage.none;

  /// 平滑窗口大小, 与桌面版默认 5 一致
  final int smoothingWindow;

  /// 角度历史 (滑动窗口)
  final List<double> _angleHistory = [];

  /// 上次计数时间戳 (毫秒)
  int _lastCountTime = 0;

  /// 最小 rep 间隔 (毫秒), 桌面版为 0.5s
  static const int _minRepIntervalMs = 500;

  ExerciseCounter({this.smoothingWindow = 5});

  /// 重置计数状态
  void reset() {
    count = 0;
    _stage = Stage.none;
    _leftLegStage = Stage.none;
    _rightLegStage = Stage.none;
    _angleHistory.clear();
    _lastCountTime = 0;
  }

  /// 三点夹角 (b 为顶点), 单位度数
  /// 返回 null 表示点无效 (含 NaN 或 (0,0))
  static double? calculateAngle(Float64List a, Float64List b, Float64List c) {
    // 检查点有效性 (与桌面版一致: 任意点为 (0,0) 视为无效)
    if ((a[0] == 0 && a[1] == 0) ||
        (b[0] == 0 && b[1] == 0) ||
        (c[0] == 0 && c[1] == 0)) {
      return null;
    }

    final baX = a[0] - b[0];
    final baY = a[1] - b[1];
    final bcX = c[0] - b[0];
    final bcY = c[1] - b[1];

    final baNorm = math.sqrt(baX * baX + baY * baY);
    final bcNorm = math.sqrt(bcX * bcX + bcY * bcY);
    if (baNorm == 0 || bcNorm == 0) {
      return null;
    }

    var cosine = (baX * bcX + baY * bcY) / (baNorm * bcNorm);
    // 裁剪到 [-1, 1] 防止数值误差导致 acos 返回 NaN
    cosine = cosine.clamp(-1.0, 1.0);
    return math.acos(cosine) * 180.0 / math.pi;
  }

  /// 对角度做滑动窗口平滑: 中位数过滤 + 2σ 离群点去除后取均值
  /// 与桌面版 smooth_angle 完全一致
  double? smoothAngle(double? angle) {
    if (angle == null) return null;

    _angleHistory.add(angle);
    if (_angleHistory.length > smoothingWindow) {
      _angleHistory.removeAt(0);
    }
    if (_angleHistory.length < 3) return angle;

    final sorted = List<double>.of(_angleHistory)..sort();
    final median = _medianOfSorted(sorted);

    final std = _stdDev(_angleHistory);
    final filtered = _angleHistory.where((v) => (v - median).abs() <= 2 * std).toList();
    return filtered.isEmpty ? angle : filtered.reduce((a, b) => a + b) / filtered.length;
  }

  /// 通用计数入口: 给定 17 个关键点与运动类型, 返回当前角度 (供 UI 显示)
  /// 内部维护 count 状态
  ///
  /// (方法名与字段 `count` 冲突, 故改用 `update` - 与 ChangeNotifier 惯例一致)
  double? update(Keypoints keypoints, ExerciseConfig config) {
    final kp = config.keypoints;

    final leftAngle = calculateAngle(
      keypoints[kp.left[0]],
      keypoints[kp.left[1]],
      keypoints[kp.left[2]],
    );
    final rightAngle = calculateAngle(
      keypoints[kp.right[0]],
      keypoints[kp.right[1]],
      keypoints[kp.right[2]],
    );

    if (leftAngle == null || rightAngle == null) return null;

    // 腿部动作走独立逻辑
    if (config.isLegExercise) {
      return _countLegExercise(leftAngle, rightAngle, config);
    }

    final avg = (leftAngle + rightAngle) / 2;
    final smoothed = smoothAngle(avg);
    if (smoothed == null) return null;

    final direction = config.inferDirection;
    final upThreshold = config.upAngle;
    final downThreshold = config.downAngle;

    if (direction == 'decreasing') {
      // 角度减小算一次 (如深蹲、俯卧撑: 起身大角度, 下蹲/下压小角度)
      if (smoothed > downThreshold) {
        _stage = Stage.down;
      } else if (smoothed < upThreshold &&
          _stage == Stage.down &&
          _checkRepTiming()) {
        _stage = Stage.up;
        count++;
        _lastCountTime = _nowMs();
      }
    } else {
      // 角度增大算一次 (如侧平举: 起始小角度, 抬起大角度)
      if (smoothed > upThreshold) {
        _stage = Stage.up;
      } else if (smoothed < downThreshold &&
          _stage == Stage.up &&
          _checkRepTiming()) {
        _stage = Stage.down;
        count++;
        _lastCountTime = _nowMs();
      }
    }
    return smoothed;
  }

  /// 腿部动作: 左右腿独立计数, 任意一条腿完成一次 up-down 循环即 +1
  double _countLegExercise(double leftAngle, double rightAngle, ExerciseConfig config) {
    final direction = config.inferDirection;
    final upThreshold = config.upAngle;
    final downThreshold = config.downAngle;

    if (_checkRepTiming()) {
      // 左腿
      _trackLeg(leftAngle, direction, upThreshold, downThreshold, (s) => _leftLegStage = s, () => _leftLegStage);
      // 右腿
      _trackLeg(rightAngle, direction, upThreshold, downThreshold, (s) => _rightLegStage = s, () => _rightLegStage);
    }
    return (leftAngle + rightAngle) / 2;
  }

  /// 单条腿的阶段跟踪与计数 (辅助 _countLegExercise)
  void _trackLeg(
    double angle,
    String direction,
    double upThreshold,
    double downThreshold,
    void Function(Stage) setStage,
    Stage Function() getStage,
  ) {
    bool ready;
    bool counted;
    Stage readyStage;
    Stage countedStage;

    if (direction == 'decreasing') {
      ready = angle > downThreshold;
      counted = angle < upThreshold && getStage() == Stage.down;
      readyStage = Stage.down;
      countedStage = Stage.up;
    } else {
      ready = angle > upThreshold;
      counted = angle < downThreshold && getStage() == Stage.up;
      readyStage = Stage.up;
      countedStage = Stage.down;
    }

    if (ready) {
      setStage(readyStage);
    } else if (counted) {
      count++;
      _lastCountTime = _nowMs();
      setStage(countedStage);
    }
  }

  bool _checkRepTiming() {
    final now = _nowMs();
    return now - _lastCountTime >= _minRepIntervalMs;
  }

  static int _nowMs() => DateTime.now().millisecondsSinceEpoch;

  static double _medianOfSorted(List<double> sorted) {
    final n = sorted.length;
    final mid = n ~/ 2;
    return n.isEven ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid];
  }

  static double _stdDev(List<double> xs) {
    if (xs.isEmpty) return 0;
    final mean = xs.reduce((a, b) => a + b) / xs.length;
    final variance = xs.map((x) => (x - mean) * (x - mean)).reduce((a, b) => a + b) / xs.length;
    return math.sqrt(variance);
  }
}
